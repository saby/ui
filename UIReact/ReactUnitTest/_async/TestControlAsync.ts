import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlAsync');

class TestControlAsync extends Control {
    protected _template: TemplateFunction = template;
}

export = TestControlAsync;
