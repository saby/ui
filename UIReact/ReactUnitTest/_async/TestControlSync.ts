import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlSync');

class TestControlSync extends Control {
    protected _template: TemplateFunction = template;
}

export = TestControlSync;
