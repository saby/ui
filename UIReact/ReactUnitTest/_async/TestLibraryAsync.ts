import { Control, TemplateFunction } from 'UI/Base';
import template = require('wml!ReactUnitTest/_async/TestControlAsync');

class ExportControl extends Control {
    protected _template: TemplateFunction = template;
}

function exportFunction(echo: string): string {
    return echo;
}
function exportSyncFunction(echo: string): string {
    return echo;
}

export { exportFunction, exportSyncFunction, ExportControl };
