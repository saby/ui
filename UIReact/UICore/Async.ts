/**
 * Библиотека абстрактной реализации контейнера для асинхронной загрузки контролов.
 * @library UICore/Async
 * @includes Async UICore/_async/Async
 */

export {
    default as Async,
    IAsyncOptions,
    TAsyncStateReceived,
} from 'UICore/_async/Async';
