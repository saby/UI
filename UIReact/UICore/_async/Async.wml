<ws:if data="{{ error }}">
   <div>{{ userErrorMessage }}</div>
</ws:if>
<ws:else data="{{ asyncLoading }}">
   <div>Идет загрузка шаблона...</div>
</ws:else>
<ws:else>
   <ws:partial template="{{ _options.content || optionsForComponent.resolvedTemplate }}" scope="{{ optionsForComponent }}"/>
</ws:else>