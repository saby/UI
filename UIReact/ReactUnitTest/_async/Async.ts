// @ts-nocheck
import { Async, TAsyncStateReceived, IAsyncOptions } from 'UICore/Async';

/**
 * Реализация класса UICore/Async:Async для тестов
 */
export default class AsyncTest extends Async {
    _beforeMount(options: IAsyncOptions): void {
        super._beforeMount(options);
    }

    _componentDidMount(): void {
        super._componentDidMount();
    }

    _beforeUpdate(opts: IAsyncOptions): void {
        super._beforeUpdate(opts);
    }

    _afterUpdate(): void {
        super._afterUpdate();
    }

    getError(): TAsyncStateReceived | void {
        return this.error;
    }

    getCurrentTemplateName(): string {
        return this.currentTemplateName;
    }

    getOptionsForComponent(): Record<string, unknown> {
        return this.optionsForComponent;
    }
}
