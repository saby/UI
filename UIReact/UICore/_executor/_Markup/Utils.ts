import {
    Attr,
    VoidTags as voidElements,
    TAttributes,
    IAttributes
} from 'UICommon/Executor';
import {CreateChildrenRef} from './Vdom/Refs/CreateChildrenRef';
import {CreateEventRef} from './Vdom/Refs/CreateEventRef';
import {convertAttributes, WasabyAttributes} from './Attributes';
import { ArrayUtils } from 'UICommon/Utils';

import { IWasabyEvent } from 'UICommon/Events';
import {AttrToDecorate} from './interfaces';
import { Control } from 'UICore/Base';
import { ChainOfRef, CreateOriginRef } from 'UICore/Ref';

import * as React from "react";

/**
 * @author Тэн В.А.
 */

export function createTagDefaultText(tag, attrs, children, attrToDecorate?, defCollection?): string {
    if (!attrToDecorate) {
        attrToDecorate = {};
    }
    if (!attrs) {
        attrs = {attributes: {}};
    }

    const mergedAttrs = Attr.processMergeAttributes(
        attrToDecorate.attributes as IAttributes,
        attrs.attributes as IAttributes
    );

    Object.keys(mergedAttrs).forEach((attrName) => {
        if (attrName.indexOf('top:') === 0) {
            const newAttrName = attrName.replace('top:', '');
            mergedAttrs[newAttrName] = mergedAttrs[newAttrName] || mergedAttrs[attrName];
            delete mergedAttrs[attrName];
        }
    });

    const mergedAttrsStr = mergedAttrs
        ? decorateAttrs(mergedAttrs, {})
        : '';
    // tslint:disable-next-line:no-bitwise
    if (~voidElements.indexOf(tag)) {
        return '<' + tag + mergedAttrsStr + ' />';
    }
    return '<' + tag + mergedAttrsStr + '>' + joinElementsDefault(children) + '</' + tag + '>';
}

export function createTagDefaultVdom<T extends HTMLElement, P extends React.HTMLAttributes<T>>(
    tagName: keyof React.ReactHTML,
    attrs: {
        attributes: P & WasabyAttributes;
        events: Record<string, IWasabyEvent[]>
    },
    children: React.ReactNode[] & {
        for: boolean
    },
    attrToDecorate: AttrToDecorate,
    __: unknown,
    control?: Control
): React.DetailedReactHTMLElement<P, T> {
    if (!attrToDecorate) {
        attrToDecorate = {attributes: {}, events: {}};
    }
    /* если события объявляется на контроле, и корневом элементе шаблона, то мы должны смержить события,
     * без этого события объявленные на контроле будут потеряны
     */
    const extractedEvents = { ...attrToDecorate.events, ...attrs.events };
    if (Object.keys(extractedEvents).length) {
        let eventsMeta;
        if (attrToDecorate.events && attrToDecorate.events.meta && Object.keys(attrToDecorate.events.meta).length) {
            eventsMeta = {...attrToDecorate.events.meta};
        }
        if (attrs.events && attrs.events.meta && Object.keys(attrs.events.meta).length) {
            eventsMeta = {...attrs.events.meta};
        }
        Object.defineProperty(extractedEvents, 'meta', {
            configurable: true,
            value: eventsMeta
        });
    }
    const eventsObject = {
        events: extractedEvents
    };
    /**
     * Объединяет атрибуты, указанные на элементе, с атрибутами, которые пришли сверху
     */
    const mergedAttrs = Attr.mergeAttrs(attrToDecorate.attributes, attrs.attributes);
    Object.keys(mergedAttrs).forEach((attrName) => {
        if (!mergedAttrs[attrName]) {
            delete mergedAttrs[attrName];
        }
    });
    const name = mergedAttrs.name;
    const originRef = attrs.attributes.ref;
    const chainOfRef = new ChainOfRef();
    const createChildrenRef = new CreateChildrenRef(control, name);
    const createEventRef = new CreateEventRef(tagName, eventsObject);
    chainOfRef.add(createChildrenRef);
    chainOfRef.add(createEventRef);
    if (originRef){
        chainOfRef.add(new CreateOriginRef(originRef));
    }

    const convertedAttributes = convertAttributes(mergedAttrs);

    /* не добавляем extractedEvents в новые пропсы на теге, т.к. реакт будет выводить ошибку о неизвестном свойстве
        https://online.sbis.ru/opendoc.html?guid=d90ec578-f610-4d93-acdd-656095591bc1
    */
    const newProps = {
        ...convertedAttributes,
        ref: chainOfRef.execute()
    };

    // Разворачиваем массив с детьми, так как в противном случае react считает, что мы отрисовываем список
    const flatChildren = ArrayUtils.flatten(children, true);
    if (flatChildren.for) {
        // если дети получены циклом - нужно вставлять их массивом, чтобы учитывались ключи
        return React.createElement<P, T>(tagName, newProps, flatChildren);
    }
    return React.createElement<P, T>(tagName, newProps, ...flatChildren);
}

export function joinElementsDefault(elements: string[]): string {
    if (Array.isArray(elements)) {
        let res = '';
        elements.forEach((element) => {
            if (Array.isArray(element)) {
                element = joinElementsDefault(element);
            }
            res += (element || '');
        });

        return res;
    } else {
        throw new Error('joinElements: elements is not array');
    }
}

function decorateAttrs(attr1: TAttributes, attr2: TAttributes): string {
    function wrapUndef(value: string): string {
        if (value === undefined || value === null) {
            return '';
        } else {
            return value;
        }
    }

    const attrToStr = (attrs: string[]): string => {
        let str = '';
        for (const attr in attrs) {
            if (attrs.hasOwnProperty(attr)) {
                str += (wrapUndef(attrs[attr]) !== '' ? ' ' + (attr + '="' + attrs[attr] + '"') : '');
            }
        }
        return str;
    };
    return attrToStr(Attr.joinAttrs(attr1, attr2));
}
