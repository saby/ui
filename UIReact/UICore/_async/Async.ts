/**
 * @kaizen_zone 4cbeec99-a5a5-48bf-8531-411cf558b0c0
 */
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import * as Library from 'WasabyLoader/Library';
import { constants } from 'Env/Env';
import { descriptor } from 'Types/entity';
import { Control } from 'UICore/Base';
import { IControlOptions, TemplateFunction } from 'UICommon/Base';
import { addPageDeps } from 'UICommon/Deps';
import { Logger } from 'UICommon/Utils';
import { CommonUtils } from 'UICommon/Executor';
import template = require('wml!UICore/_async/Async');

function generateErrorMsg(templateName: string, msg?: string): string {
    const tTemplate = `Ошибка загрузки контрола "${templateName}"`;
    const tHint =
        'Возможны следующие причины:\n\t \
                  • Ошибка в самом контроле\n\t \
                  • Долго отвечал БЛ метод в _beforeUpdate\n\t \
                  • Контрола не существует';
    return !msg ? `${tTemplate}\n${tHint}` : `${tTemplate}: ${msg}`;
}

export type TAsyncStateReceived = boolean | string;

export interface IAsyncOptions extends IControlOptions {
    templateName: string;
    templateOptions: IControlOptions;
    noBubbleEvents?: boolean;
}

/**
 * Абстрактная реализация контейнера для асинхронной загрузки контролов.
 * !Важно: нельзя использовать этот контейнер напрямую! Необходимо использовать {@link Controls/Container/Async}
 * @see Controls/Container/Async
 * Подробное описание и примеры вы можете найти <a href='/doc/platform/developmentapl/interface-development/pattern-and-practice/async-load/'>здесь</a>.
 *
 * @public
 */
export default abstract class Async extends Control<IAsyncOptions, TAsyncStateReceived> {
    /**
     * @event UICore/Async:Async#templateLoad Событие оповещения, что указанный в templateName шаблон загружен и вставлен в DOM
     */

    protected _template: TemplateFunction = template;
    protected currentTemplateName: string;
    protected optionsForComponent: Record<string, unknown> = {};
    /**
     * Флаг для того, чтобы избежать повторной загрузки шаблона, при изменении опций до окончания асинхронной загрузки
     */
    protected asyncLoading: boolean = false;
    /**
     * Флаг, о том, что произошла ошибка при загрузке модуля - чтобы не было циклической попытки загрузки
     */
    private loadingErrorOccurred: boolean = false;
    protected error: TAsyncStateReceived | void;
    protected userErrorMessage: string | void;
    protected defaultErrorMessage: string = 'У СБИС возникла проблема';
    /**
     * Флаг чтобы понимать, что был загружен контрол и вставлен на страницу -
     * т.к. после монтирования в DOM нужно будет опубликовать событие templateLoad
     * @private
     */
    private needNotifyOnLoad: boolean = false;

    protected _beforeMount(options: IAsyncOptions): Promise<void> | void {
        if (!options.templateName) {
            this.error =
                'В модуль Async передали не корректное имя шаблона (templateName=undefined|null|empty)';
            Logger.error(this.error);
            return;
        }

        if (constants.isBrowserPlatform && !ModulesLoader.isLoaded(options.templateName)) {
            return this._loadContentAsync(options.templateName, options.templateOptions);
        }

        this.error = this._loadContentSync(options.templateName, options.templateOptions);
        if (this.error) {
            this.userErrorMessage = this.defaultErrorMessage;
        }
    }

    protected _componentDidMount(): void {
        this._notifyOnLoad();
    }

    /**
     * Если можем подставить данные при изменении синхронно, то делаем это до обновления
     * @param {*} opts
     */
    protected _beforeUpdate(opts: IAsyncOptions): void {
        if (this.asyncLoading) {
            return;
        }

        if (opts.templateName === this.currentTemplateName) {
            // поменялись только опции шаблона
            this._insertComponent(
                this.optionsForComponent.resolvedTemplate,
                opts.templateOptions,
                opts.templateName
            );
            return;
        }

        if (ModulesLoader.isLoaded(opts.templateName)) {
            this._loadContentSync(opts.templateName, opts.templateOptions);
        }
    }

    _componentDidUpdate(): void {
        if (this.asyncLoading) {
            return;
        }
        if (this.loadingErrorOccurred) {
            this.loadingErrorOccurred = false;
            return;
        }
        if (this.currentTemplateName === this._options.templateName) {
            this._notifyOnLoad();
            return;
        }
        this._loadContentAsync(this._options.templateName, this._options.templateOptions);
    }

    protected _notifyOnLoad(): void {
        if (this.needNotifyOnLoad && !this.error && !this.asyncLoading) {
            this.needNotifyOnLoad = false;
            this._notify('templateLoad');
        }
    }

    protected _loadContentSync(name: string, options: IControlOptions): TAsyncStateReceived {
        const loaded = this._loadSync(name);
        if (loaded === null) {
            return generateErrorMsg(name);
        }

        this.needNotifyOnLoad = true;
        this._insertComponent(loaded, options, name);
        addPageDeps([Library.parse(name).name]);
        return false;
    }

    protected _loadSync<T = unknown>(name: string): T {
        try {
            const loaded = ModulesLoader.loadSync<T>(name);
            if (loaded) {
                return loaded;
            }
        } catch (err) {
            Logger.error(`Couldn't load module "${name}"`, err);
        }
        return null;
    }

    protected _loadContentAsync(name: string, options: IControlOptions): Promise<void> {
        this.asyncLoading = true;
        this.loadingErrorOccurred = false;

        return this._loadAsync(name).then(
            (loaded) => {
                this.asyncLoading = false;
                if (!loaded) {
                    this.loadingErrorOccurred = true;
                    this.error = generateErrorMsg(name);
                    Logger.warn(this.error);
                    this.userErrorMessage = this.defaultErrorMessage;
                    return;
                }

                this.needNotifyOnLoad = true;
                this._insertComponent(loaded, options, name);
            },
            (err) => {
                this.asyncLoading = false;
                this.loadingErrorOccurred = true;
                this._setErrorState(err, name);
            }
        );
    }

    protected _loadAsync(name: string): Promise<unknown> {
        return ModulesLoader.loadAsync(name).catch((error) => {
            Logger.error(`Couldn't load module "${name}"`, error);
            ModulesLoader.unloadSync(name);
            throw new Error(this.defaultErrorMessage);
        });
    }

    protected _setErrorState(err: Error, name: string): void {
        this.error = generateErrorMsg(name);
        this.userErrorMessage = err.message;
    }

    protected _insertComponent(tpl: unknown, options: IControlOptions, templateName: string): void {
        this.error = '';
        this.currentTemplateName = templateName;
        this.optionsForComponent = {};
        const opts = options || {};
        for (const key in opts) {
            if (opts.hasOwnProperty(key)) {
                this.optionsForComponent[key] = opts[key];
            }
        }

        if (tpl && CommonUtils.isDefaultExport(tpl)) {
            // @ts-ignore
            this.optionsForComponent.resolvedTemplate = tpl.default;
            return;
        }
        this.optionsForComponent.resolvedTemplate = tpl;
    }

    static getOptionTypes(): Record<string, unknown> {
        return {
            templateName: descriptor(String).required(),
        };
    }
}

/**
 * @name UICore/Async:Async#content
 * @cfg {String} Содержимое контейнера.
 */

/**
 * @name UICore/Async:Async#templateName
 * @cfg {String} Имя асинхронно загружаемого контрола.
 * Можно использовать только {@link /doc/platform/developmentapl/interface-development/pattern-and-practice/javascript-libraries/#_2 публичные пути библиотеки}.
 */

/**
 * @name UICore/Async:Async#templateOptions
 * @cfg {Object} Параметры содержимого контейнера Async.
 */
