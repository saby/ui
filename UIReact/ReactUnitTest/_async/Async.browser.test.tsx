/**
 * @jest-environment jsdom
 */
import { render } from 'react-dom';
import { act } from 'react-dom/test-utils';
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import { Logger } from 'UICommon/Utils';
import { default as Async } from 'ReactUnitTest/_async/Async';

// предзагрузим модули, на которых будем тестировать Async - в тестах главное протестировать факт вызова require
import TestControlSync = require('ReactUnitTest/_async/TestControlSync');
import TestControlAsync = require('ReactUnitTest/_async/TestControlAsync');
import TestLibraryAsync = require('ReactUnitTest/_async/TestLibraryAsync');

describe('UICore/Async:Async в браузере', () => {
    let container: HTMLDivElement;
    let notifyLoadSpy;

    /**
     * Эта функция существует для того, чтобы не забывать оборачивать тики в act.
     * Это нужно, чтобы реакт реагировал на изменение времени и обновлял компоненты.
     * @param duration Значение, на которое должно продвинуться время.
     */
    function tick(duration: number): void {
        act(() => {
            jest.advanceTimersByTime(duration);
        });
    }

    /**
     * Асинхронный аналог {@link tick}, отличается тем, что эта версия позволяет выполниться коллбекам промисов.
     * @param duration Значение, на которое должно продвинуться время.
     */
    async function tickAsync(duration: number): Promise<void> {
        return act(async () => {
            // в новой версии sinon есть clock.tickAsync, который в теории делает то же самое
            jest.advanceTimersByTime(duration);
            await Promise.resolve();
        });
    }

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);

        /*
        _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
        Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
         */
        jest.useFakeTimers();

        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
        jest.spyOn(Logger, 'warn').mockImplementation(jest.fn());
        jest.spyOn(Logger, 'error').mockImplementation(jest.fn());
        // заглушка для нотификации событий контрола, просто проверим факт вызова метода публикации события
        notifyLoadSpy = jest
            .spyOn(Async.prototype, '_notify')
            .mockImplementation(jest.fn());
    });

    afterEach(() => {
        jest.useRealTimers();
        container.remove();

        jest.restoreAllMocks();
    });

    it('Синхронная загрузка контрола', () => {
        const moduleName = 'ReactUnitTest/_async/TestControlSync';

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions="{}"
                />,
                container
            );
        });

        expect(instance.getError()).toBeFalsy(); // Поле с ошибкой должно быть пустым.
        expect(instance.getCurrentTemplateName()).toEqual(moduleName);
        expect(instance.getOptionsForComponent().resolvedTemplate).toEqual(
            TestControlSync
        );
    });

    it('Событие templateLoad после синхронной загрузки контрола', async () => {
        const moduleName = 'ReactUnitTest/_async/TestControlSync';

        act(() => {
            render(
                <Async templateName={moduleName} templateOptions="{}" />,
                container
            );
        });

        async function microTaskWait() {
            return Promise.resolve();
        }
        await microTaskWait();

        expect(notifyLoadSpy).toBeCalledTimes(1);
        expect(notifyLoadSpy).toBeCalledWith('templateLoad');
    });

    it('Синхронная загрузка контрола, с ошибкой', () => {
        const moduleName = 'ReactUnitTest/_async/Fail/TestControlSync';

        // заглушка для проверки факта вызова загрузки в "require"
        const loadSyncSpy = jest
            .spyOn(ModulesLoader, 'loadSync')
            .mockReturnValueOnce(null);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить синхронную загрузку контрола
        const isLoadedSpy = jest
            .spyOn(ModulesLoader, 'isLoaded')
            .mockReturnValue(true);

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions="{}"
                />,
                container
            );
        });
        tick(0);
        // удалим заглушку функции ModulesLoader.isLoaded
        isLoadedSpy.mockRestore();

        // проверим, что была попытка синхронной загрузки контрола
        expect(loadSyncSpy).toBeCalled();
        expect(loadSyncSpy).toBeCalledWith(moduleName);
        expect(instance.getError()).toMatchSnapshot();
        expect(
            instance.getOptionsForComponent().resolvedTemplate
        ).toBeUndefined();
    });

    it('Асинхронная загрузка контрола', async () => {
        const moduleName = 'ReactUnitTest/_async/TestControlAsync';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest
            .spyOn(ModulesLoader, 'loadAsync')
            .mockResolvedValueOnce(TestControlAsync);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
        const isLoadedSpy = jest
            .spyOn(ModulesLoader, 'isLoaded')
            .mockReturnValue(false);

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions="{}"
                />,
                container
            );
        });
        // удалим заглушку функции ModulesLoader.isLoaded
        isLoadedSpy.mockRestore();

        tick(0); // вызов _afterMount

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        // вызов _beforeUpdate
        await tickAsync(0);
        // вызов _afterUpdate, чтобы опубликовалось событие 'templateLoad'
        tick(0);

        expect(instance.getError()).toBeFalsy(); // Поле с ошибкой должно быть пустым.
        expect(instance.getCurrentTemplateName()).toEqual(moduleName);
        expect(instance.getOptionsForComponent().resolvedTemplate).toEqual(
            TestControlAsync
        );
        // Должен быть вызван метод публикации события "_notify"
        expect(notifyLoadSpy).toBeCalledTimes(1);
        expect(notifyLoadSpy).toBeCalledWith('templateLoad');
    });

    it('Асинхронная загрузка контрола, с ошибкой', async () => {
        const moduleName = 'ReactUnitTest/_async/Fail/TestControlAsync';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest
            .spyOn(ModulesLoader, 'loadAsync')
            .mockRejectedValueOnce(null);
        // мокаем, чтобы не было сообщения в консоли
        jest.spyOn(ModulesLoader, 'unloadSync').mockImplementationOnce(
            jest.fn()
        );

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions="{}"
                />,
                container
            );
        });

        tick(0); // вызов _afterMount

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        // вызов _beforeUpdate
        await tickAsync(0);
        // вызов _afterUpdate
        tick(0);

        expect(instance.getError()).toMatchSnapshot();
        expect(
            instance.getOptionsForComponent().resolvedTemplate
        ).toBeUndefined();
        expect(notifyLoadSpy).not.toBeCalled(); // не должно быть вызова метода публикации события "_notify"
    });

    it('Асинхронная загрузка из библиотеки', async () => {
        const moduleName =
            'ReactUnitTest/_async/TestLibraryAsync:ExportControl';
        // заглушка для проверки факта вызова загрузки в "require"
        const loadAsyncSpy = jest
            .spyOn(ModulesLoader, 'loadAsync')
            .mockResolvedValueOnce(TestLibraryAsync.ExportControl);
        // заглушка для ModulesLoader.isLoaded, чтобы проверить асинхронную загрузку контрола
        const isLoadedSpy = jest
            .spyOn(ModulesLoader, 'isLoaded')
            .mockReturnValue(false);

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions="{}"
                />,
                container
            );
        });
        tick(0);
        // удалим заглушку функции ModulesLoader.isLoaded
        isLoadedSpy.mockRestore();

        await tickAsync(0); // вызов _afterMount

        // проверим, что был вызов асинхронной загрузки контрола
        expect(loadAsyncSpy).toBeCalled();
        expect(loadAsyncSpy).toBeCalledWith(moduleName);

        // вызов _beforeUpdate
        await tickAsync(0);
        // вызов _afterUpdate
        tick(0);

        expect(instance.getError()).toBeFalsy(); // Поле с ошибкой должно быть пустым.
        expect(instance.getCurrentTemplateName()).toEqual(moduleName);
        expect(instance.getOptionsForComponent().resolvedTemplate).toEqual(
            TestLibraryAsync.ExportControl
        );
    });
});
