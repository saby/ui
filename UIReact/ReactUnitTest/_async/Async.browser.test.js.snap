// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UICore/Async:Async в браузере Асинхронная загрузка контрола, с ошибкой 1`] = `
"Ошибка загрузки контрола \\"ReactUnitTest/_async/Fail/TestControlAsync\\"
Возможны следующие причины:
	                   • Ошибка в самом контроле
	                   • Долго отвечал БЛ метод в _beforeUpdate
	                   • Контрола не существует"
`;

exports[`UICore/Async:Async в браузере Синхронная загрузка контрола, с ошибкой 1`] = `
"Ошибка загрузки контрола \\"ReactUnitTest/_async/Fail/TestControlSync\\"
Возможны следующие причины:
	                   • Ошибка в самом контроле
	                   • Долго отвечал БЛ метод в _beforeUpdate
	                   • Контрола не существует"
`;
